### PR TITLE
Fixing DoRA docs, adding to mem opt tutorial

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Fix DoRA docstring, add to the rest of the docs

```haskell
                          ooo
                         $ o$
                        o $$
              ""$$$    o" $$ oo "
          " o$"$oo$$$"o$$o$$"$$$$$ o
         $" "o$$$$$$o$$$$$$$$$$$$$$o     o
      o$"    "$$$$$$$$$$$$$$$$$$$$$$o" "oo  o
     " "     o  "$$$o   o$$$$$$$$$$$oo$$
    " $     " "o$$$$$ $$$$$$$$$$$"$$$$$$$o
  o  $       o o$$$$$"$$$$$$$$$$$o$$"""$$$$o " "
 o          o$$$$$"    "$$$$$$$$$$ "" oo $$   o $
 $  $       $$$$$  $$$oo "$$$$$$$$o o $$$o$$oo o o
o        o $$$$$oo$$$$$$o$$$$ ""$$oo$$$$$$$$"  " "o
"   o    $ ""$$$$$$$$$$$$$$  o  "$$$$$$$$$$$$   o "
"   $      "$$$$$$$$$$$$$$   "   $$$"$$$$$$$$o  o
$   o      o$"""""$$$$$$$$    oooo$$ $$$$$$$$"  "
$      o""o $$o    $$$$$$$$$$$$$$$$$ ""  o$$$   $ o
 o     " "o "$$$$  $$$$$""""""""""" $  o$$$$$"" o o
 "  " o  o$o" $$$$o   ""           o  o$$$$$"   o
  $         o$$$$$$$oo            "oo$$$$$$$"    o
  "$   o o$o $o o$$$$$"$$$$oooo$$$$$$$$$$$$$$"o$o
    "o oo  $o$"oo$$$$$o$$$$$$$$$$$$"$$$$$$$$"o$"
     "$ooo $$o$   $$$$$$$$$$$$$$$$ $$$$$$$$o"
        "" $$$$$$$$$$$$$$$$$$$$$$" """"
                         """"""
```

### CODE REVIEW

File: docs/source/tutorials/memory_optimizations.rst [hunk 1]
```diff
@@ -292,7 +294,24 @@ Secondly, parameters which control the scale of the impact of LoRA on the model:
...
+  model.lora_attn_modules=["q_proj","k_proj","v_proj"] \
+  model.lora_rank=128 \
+  model.lora_rank=256
+
+.. code-block:: yaml
+
+  model:
+    _component_: torchtune.models.llama3.lora_llama3_8b
+    apply_lora_to_mlp: True
+    model.lora_attn_modules: ["q_proj", "k_proj", "v_proj"]
```
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:312]: these should not have `model.`

File: docs/source/tutorials/memory_optimizations.rst [hunk 2]
```diff
@@ -323,18 +342,100 @@ You can finetune using QLoRA with any of our LoRA recipes, i.e. recipes with the
...
-All the rest of the LoRA parameters remain the same for QLoRA - check out the section above on :ref:`LoRA <glossary_lora>`
-to see how to configure.
+or, by modifying a config:
+
+.. code-block:: yaml
+
+  model:
+    _component_: torchtune.models.qlora_llama3_8b
+    apply_lora_to_mlp: True
+    model.lora_attn_modules: ["q_proj", "k_proj", "v_proj"]
```
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:368]: same comment

File: docs/source/tutorials/memory_optimizations.rst [hunk 3]
```diff
@@ -323,18 +342,100 @@ You can finetune using QLoRA with any of our LoRA recipes, i.e. recipes with the
...
+Since DoRA extends LoRA, the parameters for :ref:`customizing LoRA <glossary_lora>` are identical. You can also quantize the base model weights like in :ref:`glossary_qlora` by using ``quantize=True`` to reap
+even more memory savings!
+
+.. code-block:: bash
+
+  tune run lora_finetune_single_device --config llama3/8B_lora_single_device \
+  model.apply_lora_to_mlp=True \
+  model.lora_attn_modules=["q_proj","k_proj","v_proj"] \
+  model.lora_rank=128 \
+  model.lora_rank=256 \
```
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:415]: any suggestions for lower ranks for dora? also there is a repeat param

File: docs/source/tutorials/memory_optimizations.rst [hunk 4]
```diff
@@ -292,7 +294,24 @@ Secondly, parameters which control the scale of the impact of LoRA on the model:
...
-As above, these parameters are also specified under the ``model`` flag or config entry.
+As above, these parameters are also specified under the ``model`` flag or config entry:
+
+.. code-block:: bash
+
+  tune run lora_finetune_single_device --config llama3/8B_lora_single_device  \
+  model.apply_lora_to_mlp=True \
+  model.lora_attn_modules=["q_proj","k_proj","v_proj"] \
+  model.lora_rank=128 \
+  model.lora_rank=256
```
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:305]: why are there two lora_ranks?

File: docs/source/tutorials/memory_optimizations.rst [hunk 5]
```diff
@@ -21,6 +21,7 @@ To make things easy, we've summarized these components in the following table:
    ":ref:`glossary_opt_in_bwd`", "Helps reduce memory usage when using stateful optimizers, particularly when full-finetuning large models with high gradient memory usage. This is not compatible with ``gradient_accumulation_steps``, so training may slow down due to reduced model throughput."
    ":ref:`glossary_lora`", "When you want to significantly reduce the number of trainable parameters, saving gradient and optimizer memory during training, and significantly speeding up training."
    ":ref:`glossary_qlora`", "When you need even more memory savings than LoRA, at the potential cost of some training speed. Useful for very large models or limited hardware."
+   ":ref:`glossary_dora`", "Like LoRA, DoRA can provide significant memory savings and training speed-ups. DoRA may improve performance over LoRA, particularly when using small rank updates."
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply docs/source/tutorials/memory_optimizations.rst:24]: honestly not sure, according to the paper it's just straight up better

File: docs/source/tutorials/memory_optimizations.rst [hunk 6]
```diff
@@ -323,18 +342,100 @@ You can finetune using QLoRA with any of our LoRA recipes, i.e. recipes with the
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+*What's going on here?*
+
+`DoRA <https://arxiv.org/abs/2402.09353>`_ is another PEFT technique which builds on-top of LoRA by
+further decomposing the pre-trained weights into two components: magnitude and direction. The magnitude component
+is a scalar vector that adjusts the scale, while the direction component corresponds to the original LoRA decomposition and
+updates the orientation of weights.
+
+DoRA adds a small overhead to LoRA training due to the addition of the magnitude parameter, but it has been shown to
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply docs/source/tutorials/memory_optimizations.rst:386]: not 100% but there's an added parameter and extra computation so I'd say both

File: docs/source/tutorials/memory_optimizations.rst [hunk 7]
```diff
@@ -292,7 +294,24 @@ Secondly, parameters which control the scale of the impact of LoRA on the model:
...
-As above, these parameters are also specified under the ``model`` flag or config entry.
+As above, these parameters are also specified under the ``model`` flag or config entry:
+
+.. code-block:: bash
+
+  tune run lora_finetune_single_device --config llama3/8B_lora_single_device  \
+  model.apply_lora_to_mlp=True \
+  model.lora_attn_modules=["q_proj","k_proj","v_proj"] \
+  model.lora_rank=32 \
+  model.lora_rank=64
```
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:305]:
```suggestion
  model.lora_alpha=64
```

File: docs/source/tutorials/memory_optimizations.rst [hunk 8]
```diff
@@ -323,18 +342,98 @@ You can finetune using QLoRA with any of our LoRA recipes, i.e. recipes with the
...
+  model.quantize_base=True
+
+.. code-block:: yaml
+
+  model:
+    _component_: torchtune.models.lora_llama3_8b
+    apply_lora_to_mlp: True
+    lora_attn_modules: ["q_proj", "k_proj", "v_proj"]
+    lora_rank: 16
+    lora_rank: 32
```
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:424]:
```suggestion
    lora_alpha: 32
```

**Review by RdoubleA — APPROVED** [review]
Few more typos, stamping to unblock

File: docs/source/tutorials/memory_optimizations.rst (hunk 7 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/memory_optimizations.rst:305]: how am I so blind to this typo

**Pull request merged**